operator/e2e_testing/setup/k8s_clusters.go

```diff
@@ -0,0 +1,622 @@
+// /*
+// Copyright 2025 The Grove Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// */
+
+package setup
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/ai-dynamo/grove/operator/e2e_testing/utils"
+	"github.com/docker/docker/api/types/container"
+	dockerclient "github.com/docker/docker/client"
+	"github.com/k3d-io/k3d/v5/pkg/client"
+	"github.com/k3d-io/k3d/v5/pkg/config"
+	"github.com/k3d-io/k3d/v5/pkg/config/types"
+	"github.com/k3d-io/k3d/v5/pkg/config/v1alpha5"
+	k3dlogger "github.com/k3d-io/k3d/v5/pkg/logger"
+	"github.com/k3d-io/k3d/v5/pkg/runtimes"
+	k3d "github.com/k3d-io/k3d/v5/pkg/types"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	// defaultPollTimeout is the timeout for most polling conditions
+	defaultPollTimeout = 5 * time.Minute
+	// defaultPollInterval is the interval for most polling conditions
+	defaultPollInterval = 5 * time.Second
+)
+
+// NodeTaint represents a Kubernetes node taint
+type NodeTaint struct {
+	Key    string
+	Value  string
+	Effect string
+}
+
+// NodeLabel represents a Kubernetes node label with its target node filters
+type NodeLabel struct {
+	Key   string // Label key
+	Value string // Label value
+	// k3s refers to worker nodes as agent nodes
+	NodeFilters []string // Node filters (e.g., "server:*", "agent:*", "server:0", "agent:1")
+}
+
+// ClusterConfig holds configuration for creating a k3d cluster
+type ClusterConfig struct {
+	Name             string      // Name of the k3d cluster
+	ServerNodes      int         // Number of server (non-worker) nodes
```

Comment:
Nit:

More descriptive in my opinion, and this keeps the terminiology consistent with the `kind` setup we have for development as well.

```suggestion
	ControlPlaneNodes      int         // Number of control-plane nodes
```